### PR TITLE
[Feature][Support Struct] Support read struct type in FE

### PR DESCRIPTION
**Pull request by Smith-Cruise**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Link: #12339

In FE, we do lots of trick work, in order to support struct `dot` access and make it compatible with the original system.

The root cause of this situation is that in the past, the system determined the col's catalog name, database name, and table name in the Parser, which is not acceptable for struct type.

So we do the following trick methods:

1. Change `StarRocks.g4`:

If we do `SELECT a from tbl;`, `a` must refer to a column name.

If we do `SELECT  a.b from tbl;`, `a` may be a table name(`b` is col name); `a` may be a column name(`b` is struct subfield name). In this situation, we will give them may be wrong values (catalog, db, and table name, here is table name: a, col name: b) in SlotRef, and also store the original QualifiedName[a.b], so we can dereference it in ExpressionAnalyzer based on QualifiedName.
> Why do we need to set maybe wrong values for SlotRef? Because SlotRef is heavily used in our system, we do this for compatibility with the origin code.

If we do `SELECT a[10].b from tbl;`, we know that `a[10]` must be a struct type, and `b` is a subfield name, so we can create `SubfieldExpr` directly.

2. Record struct subfield access path

SlotRef is used to refer to a column name in the past, so we need to add `usedStructFieldPos` to record the subfield access path.

3. Change the expression tree structure in `SqlToScalarOperator`, because in ExpressionAnalyzer, we can't change the expression tree's logic structure, so do this change in `SqlToScalarOperator`. If SlotRef's `usedStructFieldPos` is not empty, we will add a `SubfieldOperator` on top of `ColumnRefOperator`.

**Example**
Table Definition:  [a: INT, b: STRUCT<b1: INT, b2: INT>, c:ARRAY<STRUCT<c1: INT, b2: INT>>]

`SELECT a FROM tbl;`, following original logic.

`SELECT tbl.a FROM tbl;`, following original logic.

`SELECT b.b1 FROM tbl;`: **Do trick**
[Parser:] SlotRef(table name: b, col name: b1, qualifiedName: b.b1) [table name, col name is a wrong value]
[Analyzer:] SlotRef(table name: tbl, col name: b, usedStructFieldPos: [0]) [Set correct table name and col name]
[Convert to Operator]: SubfieldOperator(fieldname: b1, children: columnRefOperator)
... do normally

`SELECT c[10].c1 FROM tbl;`
[Parser:] SubfieldExpr(table name: c1, children: ArrayElementExpr(10, children: SlotRef(c)))
... do normally

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 275 / 322 (85.40%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[937]|
|:large_blue_circle:|com/starrocks/analysis/SubfieldExpr.java|13|22|59.09%|[32, 33, 34, 46, 47, 56, 57, 58, 62]|
|:large_blue_circle:|com/starrocks/catalog/StructType.java|21|30|70.00%|[56, 68, 73, 74, 75, 76, 77, 89, 112]|
|:large_blue_circle:|com/starrocks/connector/ColumnTypeConverter.java|34|47|72.34%|[100, 315, 321, 419, 420, 421, 422, 423, 424, 425, 426, 429, 431]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java|32|36|88.89%|[39, 81, 92, 96]|
|:large_blue_circle:|com/starrocks/analysis/SlotRef.java|49|55|89.09%|[110, 111, 112, 154, 155, 373]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Field.java|40|44|90.91%|[92, 96, 126, 155]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|13|14|92.86%|[214]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/TypeDef.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ViewDefBuilder.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/QualifiedName.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/StructField.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|9|9|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/analysis/Expr.java [hunk 1]
```diff
@@ -242,6 +242,11 @@ public Type getOriginType() {
         return originType;
     }
 
+    // Used to differ from getOriginType(), return originType directly.
+    public Type getTrueOriginType() {
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/analysis/Expr.java:246]: why do we need this

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java [hunk 2]
```diff
@@ -253,7 +255,25 @@ public ScalarOperator visitSlot(SlotRef node, Context context) {
...
                 correlation.add(columnRefOperator);
             }
-            return columnRefOperator;
+
+            // If origin type is struct type, means that node contains subfield access
+            if (node.getTrueOriginType().isStructType()) {
+                Preconditions.checkArgument(node.getUsedStructFieldPos() != null, "StructType SlotRef must have" +
+                        "an non-empty usedStructFiledPos!");
+                Preconditions.checkArgument(node.getUsedStructFieldPos().size() > 0);
+                List<Integer> usedStructFieldPos = node.getUsedStructFieldPos();
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java:264]: usedStructFieldPos looks like have same function with MapType, can we unify it?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import com.starrocks.catalog.StructType;
+import com.starrocks.catalog.Type;
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class SubfieldOperator extends ScalarOperator {
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java:17]: SubfieldOperator's behavior is a little different from CollectionElementOperator. And SubfieldOperator is mapping SubFieldExpr in BE.

I think it's better to separate it.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java (hunk 2 above)
Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java:264]: That's two different functions, I know it looks the same.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/OperatorType.java [hunk 4]
```diff
@@ -99,6 +99,7 @@ public enum OperatorType {
     CLONE,
     LAMBDA_FUNCTION,
     SUBQUERY,
+    SUB_FIELD,
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/OperatorType.java:102]: `SUBFIELD`

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 275 / 322 (85.40%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[941]|
|:large_blue_circle:|com/starrocks/analysis/SubfieldExpr.java|13|22|59.09%|[32, 33, 34, 46, 47, 56, 57, 58, 62]|
|:large_blue_circle:|com/starrocks/catalog/StructType.java|21|30|70.00%|[56, 68, 73, 74, 75, 76, 77, 89, 112]|
|:large_blue_circle:|com/starrocks/connector/ColumnTypeConverter.java|34|47|72.34%|[100, 315, 321, 419, 420, 421, 422, 423, 424, 425, 426, 429, 431]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java|32|36|88.89%|[39, 81, 92, 96]|
|:large_blue_circle:|com/starrocks/analysis/SlotRef.java|49|55|89.09%|[110, 111, 112, 154, 155, 373]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Field.java|40|44|90.91%|[92, 96, 126, 155]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|13|14|92.86%|[214]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/TypeDef.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ViewDefBuilder.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/QualifiedName.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/StructField.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|9|9|100.00%|[]|

**Review by Youngwb — APPROVED** [review]

**Comment by Smith-Cruise** [comment]
run starrocks_fe_unittest

**Review by zombee0 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Smith-Cruise** [comment]
run starrocks_fe_unittest

**Comment by Smith-Cruise** [comment]
run starrocks_fe_unittest

**Review by stephen-shelby — APPROVED** [review]

**Comment by Smith-Cruise** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 277 / 323 (85.76%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[946]|
|:large_blue_circle:|com/starrocks/analysis/SubfieldExpr.java|13|22|59.09%|[34, 35, 36, 48, 49, 58, 59, 60, 64]|
|:large_blue_circle:|com/starrocks/connector/ColumnTypeConverter.java|34|47|72.34%|[100, 315, 321, 419, 420, 421, 422, 423, 424, 425, 426, 429, 431]|
|:large_blue_circle:|com/starrocks/catalog/StructType.java|22|30|73.33%|[56, 68, 73, 74, 75, 76, 77, 112]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java|32|36|88.89%|[41, 83, 94, 98]|
|:large_blue_circle:|com/starrocks/analysis/SlotRef.java|49|55|89.09%|[110, 111, 112, 154, 155, 375]|
|:large_blue_circle:|com/starrocks/sql/analyzer/Field.java|40|44|90.91%|[92, 96, 126, 155]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|13|14|92.86%|[214]|
|:large_blue_circle:|com/starrocks/sql/analyzer/SelectAnalyzer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ScalarOperatorVisitor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/TypeDef.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ViewDefBuilder.java|13|13|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/QualifiedName.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/StructField.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/Expr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AST2SQL.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|12|12|100.00%|[]|

**Comment by Smith-Cruise** [comment]
run starrocks_admit_test

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**